### PR TITLE
[WIP] Add applications

### LINKED ISSUE #481: Deploy Application API Endpoint
- [ ] Aplication Create
- [ ] Application List
- [ ] Application Details
- [ ] Tests
- [ ] OpenAPI docs

### PR DESCRIPTION
**Pull request by tarokkk**
- [x] Aplication Create
- [x] Application List
- [x] Application Details
- [x] Application Delete
- [ ] Tests
- [ ] OpenAPI docs
Fixes #481

### CODE REVIEW

File: api/applications.go [hunk 1]
```diff
@@ -37,15 +39,103 @@ type ApplicationListResponse struct {
...
+	if err != nil {
+		log.Errorf("Error getting application: %s", err.Error())
+		c.JSON(http.StatusBadRequest, components.ErrorResponse{
+			Code:    http.StatusBadRequest,
+			Message: "Error getting application",
+			Error:   err.Error(),
+		})
+		return
+	}
+	app.Deployments = deployments
```
Comment (stoader) [inline api/applications.go]: Why do we read deployments first before knowing if the application exists at all?

The flow rather should be, shouldn't it:

db.Model(organization).Related(app)

db.Model(app).Related(&app.Deployments, "Deployments")

File: api/applications.go [hunk 2]
```diff
@@ -37,15 +39,103 @@ type ApplicationListResponse struct {
...
+
+	// Get cluster for application
+	filter := make(map[string]interface{})
+	filter["organization_id"] = organization
+	filter["id"] = app.ClusterID
+	commonCluster, ok := GetCommonClusterFromFilter(c, filter)
```
Comment (stoader) [inline api/applications.go]: Not clear why we need to use `GetCommonClusterFromFilter` here.
There is already a FK constraint on `Application.ClusterId` field, so we should be able to retrieve the Cluster that the app belongs to based on `app.ClusterID` only. Once we have the cluster db record we can create the common cluster from it.



I'd add a foreign key constraint to `Application.ClusterID` and retrieve the related cluster based on the FK form db.  The FK would also

File: api/applications.go [hunk 3]
```diff
@@ -37,15 +39,103 @@ type ApplicationListResponse struct {
...
+			Message: fmt.Sprintf("Invalid id=%q", idParam),
+			Error:   err.Error(),
+		})
+		return
+	}
+	db := model.GetDB()
+	application := &model.Application{
+		ID: uint(id),
+	}
+	var deployments []*model.Deployment
```
Comment (stoader) [inline api/applications.go]: This section is similar to what's being used in other methods. Please factor out into separate function.

File: model/application.go [hunk 4]
```diff
@@ -1,53 +1,86 @@
...
+	ID             uint          `json:"id" gorm:"primary_key"`
+	CreatedAt      time.Time     `json:"createdAt"`
+	UpdatedAt      time.Time     `json:"updatedAt"`
+	DeletedAt      *time.Time    `json:"-" sql:"index"`
+	Name           string        `json:"name"`
+	CatalogName    string        `json:"catalogName"`
+	CatalogVersion string        `json:"catalogVersion"`
+	Description    string        `json:"description"`
+	Icon           string        `json:"icon"`
+	OrganizationId uint          `json:"organizationId"`
```
Comment (stoader) [inline model/application.go]: I think `OrganizationId` is redundant here as the application belongs to a cluster and the cluster to an organisation.

File: api/applications.go [hunk 5]
```diff
@@ -119,7 +209,7 @@ func CreateApplication(c *gin.Context) {
 			return
 		}
 	}
-	am := model.ApplicationModel{
+	am := model.Application{
```
Comment (baluchicken) [inline api/applications.go]: I would put the Application into Pending State here. As I see it is never on Pending state.

File: model/application.go (hunk 4 above)
Comment (tarokkk) [reply model/application.go]: This is not the case. Someone can start a Cluster and Another can deploy Applications. We can rename this to Owner if its more feasable

**Pull request closed without merging**